src/main/scala/rps/Game.scala

```diff
@@ -5,66 +5,92 @@ import scala.util.Random
 
 object Game {
 
-  // TODO this should be an enum (or Scala equivalent of TS Enum)
-  object Weapon {
-    val Rock: String = "0" // 🪨
-    val Paper: String = "1" // 📄
-    val Scissors: String = "2" // ✂️
+  sealed trait Move
+  object Move {
+    case object Rock extends Move // 🪨
+    case object Paper extends Move // 📄
+    case object Scissors extends Move // ✂️
+    val moves = List(Rock, Paper, Scissors)
+
+    // Idea is similar to a fp-ts Encoder, I dunno if it's the right place here
+    def decode(input: String): Option[Move] = {
+      input match {
+        case "0" => Some(Move.Rock)
+        case "1" => Some(Move.Paper)
+        case "2" => Some(Move.Scissors)
+        case _   => None
+      }
+    }
+    // Idea is similar to a fp-ts Decoder, I dunno if it's the right place here
+    def encode(input: Move): String = { // TODO should it also accept an Option[Move]?
+      input match {
+        case Move.Rock     => "0"
+        case Move.Paper    => "1"
+        case Move.Scissors => "2"
+        case _ => "Invalid weapon" // Is this right?? Will it be ever used?
+      }
+    }
   }
 
-  // TODO this should be an enum (or Scala equivalent of TS Enum)
-  object Winner {
-    val Draw: String = "0" // It's a draw
-    val User: String = "1" // User wins
-    val Cpu: String = "2" // Cpu wins
+  sealed trait GameResult
+  object GameResult {
+    case object Draw extends GameResult
+    case object UserWins extends GameResult
+    case object CpuWins extends GameResult
   }
 
   def play(): Unit = {
-    println(s"""Choose your weapon:
-    ${Weapon.Rock} - ${matchWeapon(Weapon.Rock)}
-    ${Weapon.Paper} - ${matchWeapon(Weapon.Paper)}
-    ${Weapon.Scissors} - ${matchWeapon(Weapon.Scissors)}
-    """)
+    val menu = Move.moves
+      .map(m => s"${Move.encode(m)} - ${matchMove(Some(m))}")
+      .reduce((acc, curr) => s"${acc}\n${curr}")
+    println("Choose your weapon:")
+    println(menu)
+
     val rawUserInput = readLine()
-    val userWeapon = rawUserInput
-    val cpuWeapon = generateComputerMove()
+    val userMove = Move.decode(rawUserInput)
+    val cpuMove = generateComputerMove()
 
-    val result = checkWinner(userWeapon, cpuWeapon)
+    val result = checkWinner(userMove, cpuMove)
 
-    println(s"You chose: ${matchWeapon(userWeapon)} | your opponent chose: ${matchWeapon(cpuWeapon)}")
-    println(announceWinner(result))
+    println(
+      s"You chose: ${matchMove(userMove)} | your opponent chose: ${matchMove(cpuMove)}"
+    )
+    println(announceResult(result))
   }
 
-  def matchWeapon(weapon: String): String = {
-    weapon match {
-      case Weapon.Rock     => "🪨 Rock"
-      case Weapon.Paper    => "📄 Paper"
-      case Weapon.Scissors => "✂️ Scissors"
-      case _               => "💩 An invalid weapon"
+  def matchMove(move: Option[Move]): String = {
+    move.orNull match {
```

Comment:
Giustamente avevi dei dubbi su `orNull`. Qui Scala è abbastanza diverso da TS, spiego un attimo qui poi vediamo a voce magari.

In pratica `null` c'è per compatibilità con Java ma è unsafe, quindi il typechecker non ti forza a gestirlo (come TS senza strict null checks). In Scala 3 opzionalmente si può renderlo safe, ma qui no.

Perciò `.orNull` in pratica non va chiamato mai, altrimenti perdi le garanzie del type system.

In pratica con `Option` la cosa normale da fare è un pattern match che considera le due possibilità: può essere `None` oppure `Some(move)`. Puoi anche fare pattern annidati quindi qui potresti fare i casi `case Some(Move.Rock) =>` ecc. e in fondo `case None =>`.  Oppure `case Some(move) =>`, a destra della freccia gestisci i tre casi con un ulteriore match su `move`, e poi sotto `case None =>`.

(Poi ci sono tanti metodi già fatti tipo `x.getOrDefault(defaultValue)` che estrae il valore o ti dà il default se `None` e così via, ma dietro le quinte sono tutti dei match così.)

Reply:
Ottima osservazione. Rifattorizzato con un `getOrElse`.